File: 1937.MaximumNumberofPointswithCost/1937.MaximumNumberofPointswithCost.cpp
/*Dynamic programming solution to leetcode 1937
 O(m * n) time complexity
 O(n) space complexity*/
class Solution {
public:
  long long maxPoints(vector<vector<int>>& points) {
    int m = points.size(), n = points[0].size();
    vector<long long> dp(n, 0);

    long long ans = 0;
    for(int i = m - 1; i >= 0; i--) {
      vector<long long> leftToRight(n), rightToLeft(n);
      leftToRight[0] = dp[0], rightToLeft[n - 1] = dp[n - 1];
      for(int j = 1; j < n; j++) {
      leftToRight[j] = max(leftToRight[j - 1] - 1, dp[j]);
      rightToLeft[n - 1 - j] = max(rightToLeft[n - j] - 1, dp[n - 1 -j]);
      }

    for (int j = 0; j < n; j++)
      dp[j] = points[i][j] + max(leftToRight[j], rightToLeft[j]);
  }

    for(int j = 0; j < n; j++)
      ans = max(ans, dp[j]);
      return ans;
  }
};

File: 1937.MaximumNumberofPointswithCost/1937.MaximumNumberofPointswithCost.py
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        m = len(points)
        n = len(points[0])

        dp = points[0]

        leftToRight = [0] * n
        rightToLeft = [0] * n

        for i in range(1, m):
            for j in range(n):
                if j == 0:
                    leftToRight[j] = dp[j]
                else:
                    leftToRight[j] = max(leftToRight[j-1]-1, dp[j])

            for j in range(n-1, -1, -1):
                if j == n-1:
                    rightToLeft[j] = dp[j]
                else:
                    rightToLeft[j] = max(rightToLeft[j+1]-1, dp[j])

            for j in range(n):
                dp[j] = points[i][j] + max(leftToRight[j], rightToLeft[j])

        return max(dp)
